File: Data Structures & Algorithms/jump-game/submission-2.cpp
class Solution {
public:
    bool helper(vector<int>& nums, int idx, unordered_map<int, bool>& tab)
    {
        if(tab.find(idx) != tab.end())
        {
            return tab[idx];
        }
        else if(idx == nums.size() - 1)
        {
            tab[idx] = true;
            return tab[idx];
        }
        else if(nums[idx] == 0)
        {
            tab[idx] = false;
            return tab[idx];
        }
        else
        {
            bool res = false;
            for(int i = 1; i < nums[idx] + 1; i++)
            {
                int next_idx = idx + i;
                res = res || helper(nums, next_idx, tab);
            }
            tab[idx] = res;
            return tab[idx];
        }
    }

    bool canJump(vector<int>& nums) {
        unordered_map<int, bool> tab;
        return helper(nums, 0, tab);
    }
};

File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        elif len(nums) == 2:
            return max(nums)
        else:
            n = len(nums)
            tab = [0] * n
            tab[n - 1] = nums[n - 1]
            tab[n - 2] = nums[n - 2]

            for i in range(n - 3, -1, -1):
                for j in range(i + 2, n):
                    tab[i] = max(tab[i], nums[i] + tab[j])
            
            return max(tab)